--- /sdcard/stable_zen/src/parser/parser.js	2026-07-16 03:47:13.695768243 +0530
+++ /sdcard/zen/src/parser/parser.js	2026-07-18 11:41:30.100712594 +0530
@@ -7,7 +7,7 @@
 import { Lexer } from "../lexer/lexer.js";
 
 export class Parser {
-  constructor(tokens, IRB, options = { preserveComments: false }) {
+  constructor(tokens, IRB, options = { preserveComments: false}) {
     this.tokens = tokens;
     this.pos = 0;
     this.IRB = IRB;
@@ -98,7 +98,7 @@
       const stmt = this.parseStatement();
       if (stmt) body.push(stmt);
 
-      //  consume statement separator
+      //  consume statement sparator
       if (this.match("NEWLINE")) {
         this.advance();
       }
@@ -110,32 +110,31 @@
   // STATEMENTS
 
   parseStatement() {
-    if (
-      this.matchKeyword("const") &&
-      this.tokens[this.pos + 1]?.type === "IDENTIFIER" &&
-      this.tokens[this.pos + 2]?.type === "IDENTIFIER"
-    ) {
-      this.advance(); // const
-      return this.node(this.parseStructVariableDeclaration(true));
-    }
-
-    if (this.match("IDENTIFIER") && this.peek("IDENTIFIER")) {
-      return this.node(this.parseStructVariableDeclaration(false));
-    }
-
+    
+    if (this.matchKeyword("const") &&
+    this.tokens[this.pos + 1]?.type === "IDENTIFIER" &&
+    this.tokens[this.pos + 2]?.type === "IDENTIFIER") {
+    this.advance(); // const
+    return this.node(this.parseStructVariableDeclaration(true));
+}
+
+if (this.match("IDENTIFIER") && this.peek("IDENTIFIER")) {
+    return this.node(this.parseStructVariableDeclaration(false));
+}
+    
     if (
       this.match("TYPE") ||
       this.matchKeyword("auto") ||
-      this.matchKeyword("reactive") ||
+      this.matchKeyword("reactive") || 
       this.matchKeyword("const")
     ) {
       return this.node(this.parseVariableDeclaration());
     }
-
+    
     if (this.options.preserveComments) {
-      if (this.match("COMMENT")) {
-        return this.parseComment();
-      }
+    if (this.match("COMMENT")) {
+      return this.parseComment();
+    }
     }
 
     if (this.matchKeyword("switch")) {
@@ -225,29 +224,40 @@
     }
 
     if (
-      this.matchKeyword("await") &&
-      this.tokens[this.pos + 1]?.type === "IDENTIFIER" &&
-      this.tokens[this.pos + 2]?.type === "LEFT_PARENTHESIS"
-    ) {
-      this.advance();
-
-      const name = this.current().value;
+  this.matchKeyword("await") &&
+  this.tokens[this.pos + 1]?.type === "IDENTIFIER" &&
+  (
+    this.tokens[this.pos + 2]?.type === "LEFT_PARENTHESIS" ||
+    this.tokens[this.pos + 2]?.value === "<"
+  )
+) {
+  this.advance();
 
-      this.advance();
+  const name = this.current().value;
 
-      return this.node(
-        this.parseCall(
-          name,
-          true, // isAwait
-        ),
-      );
-    }
+  this.advance();
 
-    if (this.match("IDENTIFIER") && this.peek("LEFT_PARENTHESIS")) {
-      const name = this.current().value;
-      this.advance();
-      return this.node(this.parseCall(name, false));
-    }
+  return this.node(
+    this.parseCall(
+      name,
+      true, // isAwait
+    ),
+  );
+}
+
+if (
+  this.match("IDENTIFIER") &&
+  (
+    this.peek("LEFT_PARENTHESIS") ||
+    this.tokens[this.pos + 1]?.value === "<"
+  )
+) {
+  
+  const name = this.current().value;
+  
+  this.advance();
+  return this.node(this.parseCall(name, false));
+}
 
     if (this.matchKeyword("continue")) {
       this.expectKeyword("continue");
@@ -293,34 +303,34 @@
       expression: expr,
     });
   }
-
+  
   parseStructVariableDeclaration(isConst = false) {
     const struct_ref = this.advance().value;
     const name = this.advance().value;
 
     let value = null;
     if (this.current().value === "=") {
-      this.advance();
-      value = this.node(this.parseExpression());
+        this.advance();
+        value = this.node(this.parseExpression());
     }
 
     return this.node({
-      type: ParserTypes.VARIABLE_DECLARATION,
-      struct_ref,
-      name,
-      value,
-      isConstant: isConst,
+        type: ParserTypes.VARIABLE_DECLARATION,
+        struct_ref,
+        name,
+        value,
+        isConstant: isConst,
     });
-  }
-
+}
+  
   parseComment() {
     const token = this.advance();
 
     return {
       type: ParserTypes.COMMENT,
-      value: token.value,
+      value: token.value
     };
-  }
+}
 
   parseMap(isConstant = false) {
     this.expectKeyword("Map");
@@ -336,7 +346,7 @@
     } else {
       // sugar: Map a => Map a = {}
       value = {
-        type: "MAP_LITERAL",
+        type: "STRUCT_LITERAL",
         properties: [],
       };
     }
@@ -411,7 +421,7 @@
     this.expect("BLOCK_END");
 
     return this.node({
-      type: ParserTypes.MAP_LITERAL,
+      type: ParserTypes.STRUCT_LITERAL,
       properties,
     });
   }
@@ -539,11 +549,7 @@
         isThreadfn = true;
         this.advance();
       } else if (this.matchKeyword("fn")) {
-        this.IRB.emitError(
-          "SyntaxError",
-          "method should be declared without 'fn' keyword",
-          this.lineAndColumn(),
-        );
+        this.IRB.emitError("SyntaxError", "method should be declared without 'fn' keyword", this.lineAndColumn());
       }
 
       // METHOD
@@ -693,8 +699,9 @@
   }
 
   parseList(isConstant = false) {
+    
     const generic = this.parseListGeneric();
-
+    
     const name = this.expect("IDENTIFIER").value;
 
     let value = null;
@@ -842,7 +849,7 @@
     }
 
     if (this.matchKeyword("fn")) {
-      return this.parseFnParam();
+      return this.parseFParam();
     }
 
     if (this.matchKeyword("Map")) {
@@ -946,20 +953,20 @@
     this.expect("RIGHT_PARENTHESIS");
 
     let returnType = {
-      type: "void",
+  type: "void",
+  dimensions: [],
+};
+
+if (!this.match("COMMA") && !this.match("RIGHT_PARENTHESIS")) {
+  if (this.matchKeyword("void")) {
+    returnType = {
+      type: this.expectKeyword("void").value,
       dimensions: [],
     };
-
-    if (!this.match("COMMA") && !this.match("RIGHT_PARENTHESIS")) {
-      if (this.matchKeyword("void")) {
-        returnType = {
-          type: this.expectKeyword("void").value,
-          dimensions: [],
-        };
-      } else {
-        returnType = this.parseType(true);
-      }
-    }
+  } else {
+    returnType = this.parseType(true);
+  }
+}
 
     return {
       type: "Function",
@@ -985,7 +992,7 @@
       } else if (this.matchKeyword("extern")) this.advance();
 
       this.expectKeyword("fn");
-    }
+    } 
 
     let name;
 
@@ -1008,23 +1015,23 @@
 
     while (!this.match("RIGHT_PARENTHESIS")) {
       this.skipNewlines();
-
+      
       let isConstant = false;
 
-      if (this.matchKeyword("const")) {
+    if (this.matchKeyword("const")) {
         isConstant = true;
         this.advance();
-      }
+    }
 
       const t = this.parseType();
-
+      
       if (isConstant && t.type === "Function") {
-        this.IRB.emitError(
-          "TypeError",
-          "function callback parameters cannot be declared as 'const'",
-          this.lineAndColumn(),
-        );
-      }
+  this.IRB.emitError(
+    "TypeError",
+    "function callback parameters cannot be declared as 'const'",
+    this.lineAndColumn(),
+  );
+}
 
       let name;
 
@@ -1034,16 +1041,6 @@
         name = this.expect("IDENTIFIER").value;
       }
 
-      // Zen v1 limitation:
-      // Map parameters are temporarily disabled.
-
-      if (t.type === "Map") {
-        this.IRB.emitError(
-          "SemanticError",
-          "Map values cannot be used as function parameters.",
-        );
-      }
-
       this.skipNewlines();
       let isRest = false;
 
@@ -1207,32 +1204,6 @@
       });
     }
 
-    // LOOP IN (MAP)
-    // loop (key in map)
-    /*
-    if (isLoopIn) {
-      
-      const keyName = this.expect("IDENTIFIER").value;
-      
-      this.expectKeyword("in");
-      
-      const iterable = this.parseExpression();
-      this.expect("RIGHT_PARENTHESIS");
-      
-      const body =
-        this.match("BLOCK_START") ?
-        this.parseBlock() :
-        this.parseStatement();
-      
-      return this.node({
-        type: ParserTypes.LOOP_IN,
-        keyName,
-        iterable,
-        body
-      });
-    }
-    */
-
     // loop (init, condition, update)
 
     let first;
@@ -1366,7 +1337,7 @@
 
   parseVariableDeclaration() {
     let isConst = false;
-
+    
     if (this.match("KEYWORD")) {
       const keyVal = this.expect("KEYWORD").value;
 
@@ -1374,30 +1345,30 @@
         isConst = true;
       }
     }
-
+    
     let haveReactive = false;
     if (this.matchKeyword("reactive")) {
       haveReactive = true;
       this.advance();
     }
-
+    
     if (haveReactive && isConst) {
-      this.IRB.emitError(
+    this.IRB.emitError(
         "SyntaxError",
         "variable cannot be both 'const' and 'reactive'",
         this.lineAndColumn(),
-      );
-    }
+    );
+}
 
     // redirect to List or Map so const works there too.
     if (this.current().value === "List") {
-      return this.parseList(isConst);
-    }
-
-    if (this.current().value === "Map") {
-      return this.parseMap(isConst);
-    }
+    return this.parseList(isConst);
+}
 
+if (this.current().value === "Map") {
+    return this.parseMap(isConst);
+}
+    
     const dataType = this.advance().value;
 
     const name = this.expect("IDENTIFIER").value;
@@ -1764,9 +1735,16 @@
 
         continue;
       }
+      
+      if (this.current()?.value === "<") {
+        this.advance(); // <
+    expr.generic = this.parseListGeneric();
+    this.advance(); // >
+}
 
       if (this.match("LEFT_PARENTHESIS")) {
         this.advance();
+        
 
         const args = [];
 
@@ -1795,6 +1773,7 @@
 
         expr = {
           type: ParserTypes.CALL,
+          generic: expr.generic,
           callee: expr,
           isAwait,
           args,
@@ -1917,11 +1896,42 @@
     // VARIABLE
     if (token.type === "IDENTIFIER") {
       this.advance();
+      
+      if (
+  this.matchKeyword("await") &&
+  this.tokens[this.pos + 1]?.type === "IDENTIFIER" &&
+  (
+    this.tokens[this.pos + 2]?.type === "LEFT_PARENTHESIS" ||
+    this.tokens[this.pos + 2]?.value === "<"
+  )
+) {
+  this.advance();
 
-      // function call
-      if (this.match("LEFT_PARENTHESIS")) {
-        return this.parseCall(token.value, false);
-      }
+  const name = this.current().value;
+
+  this.advance();
+
+  return this.node(
+    this.parseCall(
+      name,
+      true, // isAwait
+    ),
+  );
+}
+
+if (
+  this.match("IDENTIFIER") &&
+  (
+    this.peek("LEFT_PARENTHESIS") ||
+    this.tokens[this.pos + 1]?.value === "<"
+  )
+) {
+  const name = this.current().value;
+  this.advance();
+  return this.node(this.parseCall(name, false));
+}
+      
+      
 
       return this.node({
         type: ParserTypes.VARIABLE,
@@ -1993,6 +2003,25 @@
   // FUNCTION CALL
 
   parseCall(name, isAwait = false) {
+    
+    let generic = null;
+
+if (this.current().value === "<") {
+  this.advance();
+
+  generic = this.parseListGeneric();
+
+  if (this.current().value === ">") {
+    this.advance();
+  } else {
+    this.IRB.emitError(
+      "SyntaxError",
+      "Expected '>' after generic type",
+      this.lineAndColumn(),
+    );
+  }
+}
+    
     this.expect("LEFT_PARENTHESIS");
     this.skipNewlines();
     const args = [];
@@ -2016,6 +2045,7 @@
       isInbuilt: BUILTIN_FUNCTIONS.includes(name),
       type: ParserTypes.CALL,
       name,
+      generic,
       isAwait,
       args,
     });
